cmd/kubectl-gadget/deploy.go

```diff
@@ -423,6 +424,8 @@ func runDeploy(cmd *cobra.Command, args []string) error {
 					gadgetContainer.Env[i].Value = runtimesConfig.Docker
 				case utils.GadgetEnvironmentPodmanSocketpath:
 					gadgetContainer.Env[i].Value = runtimesConfig.Podman
+				case experimental.EnvName:
+					gadgetContainer.Env[i].Value = strconv.FormatBool(experimental.Enabled())
```

Comment:
What about making it configurable through a flag `--experimental` as well?

Reply:
I agree it is not the cleanest solution, but that's how many tools already work. For instance, [kubectl](https://github.com/kubernetes/client-go/blob/92512ee2b8cf6696e9909245624175b7f0c971d9/tools/clientcmd/loader.go#L160-L216) allows setting the kubeconfig file through `--kubeconf` or env variable `KUBECONF`, otherwise it will use the default path: `$HOME/.kube/config`. That's usually the precedence order most of the tools use. BTW, that's the same precedence used in [`kubectl-aks`](https://github.com/Azure/kubectl-aks/blob/main/docs/config.md#precedence-of-configuration) 🙂.

All this is because I find it weird to set all the configurations through flags except for this one:

```bash
IG_EXPERIMENTAL=true kubectl gadget deploy --hook-mode="podinformer" --image "myImage" --image-pull-policy="never"
```

However, given the nature of this specific flag: "experimental", I'm fine leaving it to use only the env variable.

Reply:
I also think it would be better to have the `--experimental` flag to actually deploy IG with experimental features enabled - as it can be documented more cleanly in the deploy command. `IG_EXPERIMENTAL` on the other hand should IMHO apply directly to (in this case) `kubectl-gadget deploy` in that it would show/enable any experimental features regarding the deploy command, but not deploy an IG with the experimental features enabled.

Maybe to separate and clarify it even more, we should call the flag for the deployment `--enable-experimental-features`.

Reply:
> [kubectl](https://github.com/kubernetes/client-go/blob/92512ee2b8cf6696e9909245624175b7f0c971d9/tools/clientcmd/loader.go#L160-L216) allows setting the kubeconfig file through --kubeconf or env variable KUBECONF

That works fine in that case because you can use `--kubeconfig` or `KUBECONF` for all possible commands. In the case of `kubectl gadget` `--experimental` will only be available for the deploy command, all other commands will need `IG_EXPERIMENTAL`, and that's confusing IMO. 

> All this is because I find it weird to set all the configurations through flags except for this one:

Please notice that `IG_EXPERIMENTAL` applies to the whole kubectl-gadget, not only to the deploy command.

> IG_EXPERIMENTAL on the other hand should IMHO apply directly to (in this case) kubectl-gadget deploy in that it would show/enable any experimental features regarding the deploy command, but not deploy an IG with the experimental features enabled.

I don't see why we should differentiate the experimental features in the CLI and in ig-k8s. IMO, if you use `IG_EXPERIMENTAL=true kubectl gadget deploy` it should set the `IG_EXPERIMENTAL` env variable in the pod enabling the experimental features that are controlled by that variable. 

It's possible to define experimental features that are controlled by an additional flag, `IG_EXPERIMENTAL=true kubectl gadget deploy --enable-my-foo-feature`, "my foo feature" will be only enabled if both, `IG_EXPERIMENTAL` & `--enable-my-foo-feature` are passed, `--enable-my-foo-feature` won't be shown if `IG_EXPERIMENTAL` isn't enabled. It's up to the developer in this case how to expose those experimental features to the user, i.e, whether to require a CLI flag or not. 

My main worry about having an `--experimental` flag for the deploy command is the following case. Let's suppose we implement a feature that requires support in both, ig-k8s and kubectl-gadget, like adding a column on a gadget: https://github.com/inspektor-gadget/inspektor-gadget/pull/1700. 

```bash
# user deploys IG enabling experimental features
$ kubectl gadget deploy --experimental

# users tries to use a gadget expecting experimental features are enabled. 
$ kubectl gadget trace exec
 # experimental column is not shown

$ kubectl gadget trace exec -o columns=my-experimental-column
"my-experimental-column is not valid"

$ kubectl gadget trace exec --experimental
"--experimental is not a valid flag"

# after some failed attempts, user realizes IG_EXPERIMENTAL has to be defined
$ IG_EXPERIMENTAL=true kubectl gadget trace exec 
# works fine 
```

On the other hand, if we only expose this through `IG_EXPERIMENTAL`:

```bash
# user enables experimental features 
$ export IG_EXPERIMENTAL=true

# user deploys IG. Experimental features are on
$ kubectl gadget deploy 

# works fine on first attempt because IG_EXPERIMENTAL is already defined. 
$ kubectl gadget trace exec 
# works fine 
```

Finally, if we ever get a configuration file for kubectl-gadget, like, `KUBECONFIG` for kubectl, we could enable that feature there without having the user to set `--experimental` during deploying.

Reply:
> # As the env variable is set, user can get the experimental column without problem

This actually depends on how the "experimental column" is implemented. If this is a new field on the event, then it'll work fine, but if this is a virtual column that is only added in experimental mode, then it won't work as the env variable is not true in the kubectl-gadget context.


> I find very odd than people using kubectl-gadget needs to set the variable in their local environment to influence the remote behavior.

Perhaps a new column is not the perfect example in this case. What I wanted to say is that there are some features that require changes in both, client and server, in those cases the "experimental mode" has to be enabled in both of them.

Reply:
> This enables all experimental features. The experimental status is tied to the deployed instance:

That's an option, but it'll require to save the state of this in a local config file or something like that. I really wanted to keep this as simple as possible but I see there are different opinions regarding this one.

Reply:
> but if this is a virtual column that is only added in experimental mode, then it won't work as the env variable is not true in the kubectl-gadget context.

> That's an option, but it'll require to save the state of this in a local config file or something like that

**Didn't think it through** but I was wondering if that experimental flag could be retrieved and saved into the `catalog.json`

Reply:
Ok folks, I went the long way and implemented this by using the `--experimental` flag on the deploy command. This PR creates a new file to store the state of the deployment: catalog + other configuration (including experimental flag). Hence, after deploying with `--experimental´, the user doesn't have to specify this variable anymore as the value is read from the "deploy state file".